File: Python/_1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses/Solution.py
# coding=utf-8
# Date: 2021/5/26 10:28
from typing import List


class Solution:
    def reverseParentheses(self, s: str) -> str:
        stk = []
        res = ""
        for c in s:
            if c == '(':
                stk.append(res)
                res = ""
            elif c == ')':
                res = stk[len(stk) - 1] + res[::-1]
                stk.pop(len(stk) - 1)
            else:
                res += c
        return res


if __name__ == "__main__":
    print(Solution().reverseParentheses("(u(love)i)"))
    print(Solution().reverseParentheses("(a(bc(def)g)h)"))
    print(Solution().reverseParentheses("a(bcdefghijkl(mno)p)q"))

File: Java/src/_1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses/Solution.java
package _1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Date: 2021/5/26 10:16
 * 执行用时：2 ms, 在所有 Java 提交中击败了70.99%的用户
 * 内存消耗：36.6 MB, 在所有 Java 提交中击败了85.15%的用户
 */
public class Solution {

    public String reverseParentheses(String s) {
        Deque<StringBuilder> stk = new ArrayDeque<>();
        StringBuilder res = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (c == '(') {
                stk.push(res);
                res = new StringBuilder();
            } else if (c == ')') {
                res.reverse();
                res = stk.pop().append(res);
            } else res.append(c);
        }
        return res.toString();
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).reverseParentheses("(u(love)i)"));
        System.out.println((new Solution()).reverseParentheses("(a(bc(def)g)h)"));
        System.out.println((new Solution()).reverseParentheses("a(bcdefghijkl(mno)p)q"));
    }
}
